File: test/integration.test.ts
import test from "node:test";
import assert from "node:assert/strict";
import {spawnSync} from "node:child_process";
import {mkdtempSync, mkdirSync, rmSync, writeFileSync} from "node:fs";
import {tmpdir} from "node:os";
import path from "node:path";
import {discoverRepos} from "../src/discover.ts";
import {fastForward, getShortStatus} from "../src/git.ts";
import {canFastForward, parseStatus} from "../src/status.ts";

function gitAvailable(): boolean {
  return spawnSync("git", ["--version"], {encoding: "utf8"}).status === 0;
}

function runGit(cwd: string, args: readonly string[]): void {
  const result = spawnSync("git", ["-C", cwd, ...args], {encoding: "utf8"});
  assert.equal(result.status, 0, result.stderr || result.stdout);
}

function withTempDir(fn: (directory: string) => void): void {
  const directory = mkdtempSync(path.join(tmpdir(), "gitsy-"));
  try {
    fn(directory);
  } finally {
    rmSync(directory, {recursive: true, force: true});
  }
}

test("discovers clean child repos and filters status by changed flag", {skip: !gitAvailable()}, () => {
  withTempDir((directory) => {
    const repo = path.join(directory, "repo");
    mkdirSync(repo);
    runGit(repo, ["init"]);

    const repos = discoverRepos({cwd: directory, maxDepth: 3});
    assert.equal(repos.length, 1);
    assert.equal(repos[0]?.displayName, "repo");

    const cleanStatus = parseStatus(getShortStatus(repo).stdout);
    assert.equal(cleanStatus.changed, false);

    writeFileSync(path.join(repo, "README.md"), "hello\n");
    const dirtyStatus = parseStatus(getShortStatus(repo).stdout);
    assert.equal(dirtyStatus.changed, true);
  });
});

test("discovers linked worktrees from a discovered repo", {skip: !gitAvailable()}, () => {
  withTempDir((directory) => {
    const repo = path.join(directory, "repo");
    const worktree = path.join(directory, "linked-worktree");
    mkdirSync(repo);
    runGit(repo, ["init"]);
    runGit(repo, ["config", "user.email", "gitsy@example.com"]);
    runGit(repo, ["config", "user.name", "Gitsy Test"]);
    writeFileSync(path.join(repo, "README.md"), "hello\n");
    runGit(repo, ["add", "README.md"]);
    runGit(repo, ["commit", "-m", "initial"]);
    runGit(repo, ["worktree", "add", "-b", "feature", worktree]);

    const repos = discoverRepos({cwd: directory, maxDepth: 3});
    const names = repos.map((discovered) => discovered.displayName).sort();

    assert.deepEqual(names, ["linked-worktree", "repo"]);
  });
});

test("CLI --no-fetch renders repo status without fetching", {skip: !gitAvailable()}, () => {
  withTempDir((directory) => {
    const repo = path.join(directory, "repo");
    mkdirSync(repo);
    runGit(repo, ["init"]);
    writeFileSync(path.join(repo, "README.md"), "hello\n");

    const cliPath = path.resolve(import.meta.dirname, "..", "src", "index.ts");
    const result = spawnSync("node", [cliPath, "--no-fetch"], {
      cwd: directory,
      encoding: "utf8",
      env: {...process.env, FORCE_COLOR: "0", CI: "true"},
    });

    assert.equal(result.status, 0, result.stderr || result.stdout);
    assert.ok(result.stdout.includes("repo"), `Expected output to include repo name: ${result.stdout}`);
  });
});

test("--sync safely fast-forwards a clone that is strictly behind", {skip: !gitAvailable()}, () => {
  withTempDir((directory) => {
    const origin = path.join(directory, "origin");
    const clone = path.join(directory, "clone");
    mkdirSync(origin);
    runGit(origin, ["init"]);
    runGit(origin, ["config", "user.email", "gitsy@example.com"]);
    runGit(origin, ["config", "user.name", "Gitsy Test"]);
    writeFileSync(path.join(origin, "README.md"), "one\n");
    runGit(origin, ["add", "README.md"]);
    runGit(origin, ["commit", "-m", "first"]);

    runGit(directory, ["clone", origin, clone]);
    runGit(clone, ["config", "user.email", "gitsy@example.com"]);
    runGit(clone, ["config", "user.name", "Gitsy Test"]);

    writeFileSync(path.join(origin, "README.md"), "one\ntwo\n");
    runGit(origin, ["add", "README.md"]);
    runGit(origin, ["commit", "-m", "second"]);

    runGit(clone, ["fetch"]);
    const before = parseStatus(getShortStatus(clone).stdout);
    assert.equal(before.branch?.behind, 1);
    assert.equal(before.branch?.ahead, 0);
    assert.equal(canFastForward(before), true);

    const ff = fastForward(clone);
    assert.equal(ff.ok, true, ff.stderr);

    const after = parseStatus(getShortStatus(clone).stdout);
    assert.equal(after.branch?.behind, 0);
    assert.equal(canFastForward(after), false);
  });
});

test("--sync refuses to fast-forward a diverged clone", {skip: !gitAvailable()}, () => {
  withTempDir((directory) => {
    const origin = path.join(directory, "origin");
    const clone = path.join(directory, "clone");
    mkdirSync(origin);
    runGit(origin, ["init"]);
    runGit(origin, ["config", "user.email", "gitsy@example.com"]);
    runGit(origin, ["config", "user.name", "Gitsy Test"]);
    writeFileSync(path.join(origin, "README.md"), "one\n");
    runGit(origin, ["add", "README.md"]);
    runGit(origin, ["commit", "-m", "first"]);

    runGit(directory, ["clone", origin, clone]);
    runGit(clone, ["config", "user.email", "gitsy@example.com"]);
    runGit(clone, ["config", "user.name", "Gitsy Test"]);

    writeFileSync(path.join(origin, "README.md"), "one\norigin\n");
    runGit(origin, ["add", "README.md"]);
    runGit(origin, ["commit", "-m", "origin change"]);

    writeFileSync(path.join(clone, "LOCAL.md"), "local\n");
    runGit(clone, ["add", "LOCAL.md"]);
    runGit(clone, ["commit", "-m", "local change"]);

    runGit(clone, ["fetch"]);
    const status = parseStatus(getShortStatus(clone).stdout);
    assert.equal(status.branch?.ahead, 1);
    assert.equal(status.branch?.behind, 1);
    assert.equal(canFastForward(status), false);

    const ff = fastForward(clone);
    assert.equal(ff.ok, false);
  });
});

File: integration_test.go
package gitsy_test

import (
	"os"
	"os/exec"
	"path/filepath"
	"reflect"
	"strings"
	"testing"

	"github.com/flexdinesh/gitsy/internal/discover"
	"github.com/flexdinesh/gitsy/internal/git"
	"github.com/flexdinesh/gitsy/internal/status"
)

func TestDiscoversCleanChildReposAndFiltersStatusByChangedFlag(t *testing.T) {
	requireGit(t)
	dir := t.TempDir()
	repo := filepath.Join(dir, "repo")
	mkdirAll(t, repo)
	runGit(t, repo, "init")

	repos := discover.Discover(discover.Options{Cwd: dir, MaxDepth: 3})
	if len(repos) != 1 {
		t.Fatalf("expected one repo, got %d", len(repos))
	}
	if repos[0].DisplayName != "repo" {
		t.Fatalf("expected display name repo, got %s", repos[0].DisplayName)
	}

	cleanStatus := status.Parse(git.ShortStatus(repo).Stdout)
	if cleanStatus.Changed {
		t.Fatal("expected clean repo to be unchanged")
	}

	writeFile(t, filepath.Join(repo, "README.md"), "hello\n")
	dirtyStatus := status.Parse(git.ShortStatus(repo).Stdout)
	if !dirtyStatus.Changed {
		t.Fatal("expected dirty repo to be changed")
	}
}

func TestDiscoversLinkedWorktreesFromDiscoveredRepo(t *testing.T) {
	requireGit(t)
	dir := t.TempDir()
	repo := filepath.Join(dir, "repo")
	worktree := filepath.Join(dir, "linked-worktree")
	mkdirAll(t, repo)
	runGit(t, repo, "init")
	configureGitUser(t, repo)
	writeFile(t, filepath.Join(repo, "README.md"), "hello\n")
	runGit(t, repo, "add", "README.md")
	runGit(t, repo, "commit", "-m", "initial")
	runGit(t, repo, "worktree", "add", "-b", "feature", worktree)

	repos := discover.Discover(discover.Options{Cwd: dir, MaxDepth: 3})
	names := []string{}
	for _, repo := range repos {
		names = append(names, repo.DisplayName)
	}
	if !reflect.DeepEqual(names, []string{"linked-worktree", "repo"}) {
		t.Fatalf("expected linked-worktree and repo, got %#v", names)
	}
}

func TestCLINoFetchRendersRepoStatus(t *testing.T) {
	requireGit(t)
	dir := t.TempDir()
	repo := filepath.Join(dir, "repo")
	mkdirAll(t, repo)
	runGit(t, repo, "init")
	writeFile(t, filepath.Join(repo, "README.md"), "hello\n")

	cmd := exec.Command("go", "run", filepath.Join(projectRoot(t), "cmd", "gitsy"), "--no-fetch")
	cmd.Env = append(os.Environ(), "FORCE_COLOR=0", "CI=true")
	cmd.Dir = dir
	result, err := cmd.CombinedOutput()
	if err != nil {
		t.Fatalf("go run failed: %v\n%s", err, result)
	}
	if !strings.Contains(string(result), "repo") {
		t.Fatalf("expected output to include repo name: %s", result)
	}
}

func TestSyncSafelyFastForwardsStrictlyBehindClone(t *testing.T) {
	requireGit(t)
	dir := t.TempDir()
	origin := filepath.Join(dir, "origin")
	clone := filepath.Join(dir, "clone")
	mkdirAll(t, origin)
	runGit(t, origin, "init")
	configureGitUser(t, origin)
	writeFile(t, filepath.Join(origin, "README.md"), "one\n")
	runGit(t, origin, "add", "README.md")
	runGit(t, origin, "commit", "-m", "first")

	runGit(t, dir, "clone", origin, clone)
	configureGitUser(t, clone)

	writeFile(t, filepath.Join(origin, "README.md"), "one\ntwo\n")
	runGit(t, origin, "add", "README.md")
	runGit(t, origin, "commit", "-m", "second")

	runGit(t, clone, "fetch")
	before := status.Parse(git.ShortStatus(clone).Stdout)
	if before.Branch == nil || before.Branch.Behind != 1 || before.Branch.Ahead != 0 || !status.CanFastForward(before) {
		t.Fatalf("expected clone to be fast-forwardable: %+v", before.Branch)
	}

	ff := git.FastForward(clone)
	if !ff.OK {
		t.Fatalf("expected fast-forward to succeed: %s", ff.Stderr)
	}

	after := status.Parse(git.ShortStatus(clone).Stdout)
	if after.Branch == nil || after.Branch.Behind != 0 || status.CanFastForward(after) {
		t.Fatalf("expected clone to be up to date: %+v", after.Branch)
	}
}

func TestSyncRefusesDivergedClone(t *testing.T) {
	requireGit(t)
	dir := t.TempDir()
	origin := filepath.Join(dir, "origin")
	clone := filepath.Join(dir, "clone")
	mkdirAll(t, origin)
	runGit(t, origin, "init")
	configureGitUser(t, origin)
	writeFile(t, filepath.Join(origin, "README.md"), "one\n")
	runGit(t, origin, "add", "README.md")
	runGit(t, origin, "commit", "-m", "first")

	runGit(t, dir, "clone", origin, clone)
	configureGitUser(t, clone)

	writeFile(t, filepath.Join(origin, "README.md"), "one\norigin\n")
	runGit(t, origin, "add", "README.md")
	runGit(t, origin, "commit", "-m", "origin change")

	writeFile(t, filepath.Join(clone, "LOCAL.md"), "local\n")
	runGit(t, clone, "add", "LOCAL.md")
	runGit(t, clone, "commit", "-m", "local change")

	runGit(t, clone, "fetch")
	parsed := status.Parse(git.ShortStatus(clone).Stdout)
	if parsed.Branch == nil || parsed.Branch.Ahead != 1 || parsed.Branch.Behind != 1 || status.CanFastForward(parsed) {
		t.Fatalf("expected diverged clone not to be fast-forwardable: %+v", parsed.Branch)
	}

	ff := git.FastForward(clone)
	if ff.OK {
		t.Fatal("expected fast-forward to fail")
	}
}

func requireGit(t *testing.T) {
	t.Helper()
	if err := exec.Command("git", "--version").Run(); err != nil {
		t.Skip("git is not available")
	}
}

func runGit(t *testing.T, cwd string, args ...string) {
	t.Helper()
	result := exec.Command("git", append([]string{"-C", cwd}, args...)...)
	output, err := result.CombinedOutput()
	if err != nil {
		t.Fatalf("git -C %s %s failed: %v\n%s", cwd, strings.Join(args, " "), err, output)
	}
}

func configureGitUser(t *testing.T, repo string) {
	t.Helper()
	runGit(t, repo, "config", "user.email", "gitsy@example.com")
	runGit(t, repo, "config", "user.name", "Gitsy Test")
}

func mkdirAll(t *testing.T, path string) {
	t.Helper()
	if err := os.MkdirAll(path, 0o755); err != nil {
		t.Fatal(err)
	}
}

func writeFile(t *testing.T, path string, content string) {
	t.Helper()
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		t.Fatal(err)
	}
}

func projectRoot(t *testing.T) string {
	t.Helper()
	dir, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	return dir
}
